src/Paprika/Merkle/MerkleNode.cs

```diff
@@ -0,0 +1,271 @@
+using System.Buffers.Binary;
+using System.Runtime.InteropServices;
+using Paprika.Crypto;
+using Paprika.Data;
+
+namespace Paprika.Merkle;
+
+public static class Node
+{
+    public enum Type : byte
+    {
+        Leaf,
+        Extension,
+        Branch,
+    }
+
+    public static ReadOnlySpan<byte> ReadFrom(ReadOnlySpan<byte> source, out Type nodeType, out Leaf leaf, out Extension extension, out Branch branch)
+    {
+        leaf = new Leaf();
+        extension = new Extension();
+        branch = new Branch();
+
+        // TODO: It would be nice to not read the header twice:
+        // - Once to get the header
+        // - Again to read each Node type
+        Header.ReadFrom(source, out var header);
+        switch (header.NodeType)
+        {
+            case Type.Leaf:
+                nodeType = Type.Leaf;
+                return Leaf.ReadFrom(source, out leaf);
+            case Type.Extension:
+                nodeType = Type.Extension;
+                return Extension.ReadFrom(source, out extension);
+            case Type.Branch:
+                nodeType = Type.Branch;
+                return Branch.ReadFrom(source, out branch);
+            default:
+                throw new ArgumentOutOfRangeException($"Could not decode {nameof(Header)}");
+        }
+    }
+
+    private static void ValidateHeaderNodeType(Header header, Type expected)
+    {
+        if (header.NodeType != expected)
+        {
+            throw new ArgumentException($"Expected Header with {nameof(Type)} {expected}, got {header.NodeType}");
+        }
+    }
+
+    [StructLayout(LayoutKind.Explicit, Pack = 1, Size = Size)]
+    public readonly struct Header
+    {
+        public const int Size = sizeof(byte);
+
+        private const byte IsDirtyMask = 0b0001;
+        private const byte NodeTypeMask = 0b0110;
+
+        [FieldOffset(0)]
+        private readonly byte _header;
+
+        public bool IsDirty => (_header & IsDirtyMask) != 0;
+        public Type NodeType => (Type)((_header & NodeTypeMask) >> 1);
```

Comment:
Magic numbers here 😛 what is `1` ? Please extract to a `const` similar to masks.

Reply:
Changed `1` to use the `IsDirtyMask`